src/lib/HST/Frontend/FromGHC.hs

```diff
@@ -0,0 +1,395 @@
+transformTupleArg _ = error "Missing argument in tuple"
+
+-- | Transforms a GHC located pattern into an HST pattern.
+transformPat :: GHC.LPat GHC.GhcPs -> S.Pat GHC
+transformPat (GHC.L s (GHC.VarPat _ name)) =
+  S.PVar (transformSrcSpan s) (transformRdrNameUnqual name)
+transformPat (GHC.L s (GHC.ConPatIn name cpds)) =
+  let s'             = transformSrcSpan s
+      (name', isCon) = transformRdrName name
+  in  case (cpds, isCon) of
+        (GHC.InfixCon pat1 pat2, True) ->
+          S.PInfixApp s' (transformPat pat1) name' (transformPat pat2)
+        (GHC.PrefixCon pats, True) -> S.PApp s' name' (map transformPat pats)
+        (_, True) -> error "Record constructors are not supported"
+        _ -> error "Only constructors can be applied in patterns"
+-- TODO The documentation also mentions a more complicated ConPatOut.
+-- Do we need to consider that?
+transformPat (GHC.L s (GHC.TuplePat _ pats boxity)) =
+  S.PTuple (transformSrcSpan s) (transformBoxity boxity) (map transformPat pats)
+transformPat (GHC.L s (GHC.ParPat _ pat)) =
+  S.PParen (transformSrcSpan s) (transformPat pat)
+transformPat (GHC.L s (GHC.ListPat _ pats)) =
+  S.PList (transformSrcSpan s) (map transformPat pats)
+transformPat (GHC.L s (GHC.WildPat _)) = S.PWildCard (transformSrcSpan s)
+transformPat _                         = error "Unsupported pattern"
+
+-- | Transforms a GHC module name with an HST source span into an HST module
+--   name.
+transformModuleName :: S.SrcSpan GHC -> GHC.ModuleName -> S.ModuleName GHC
+transformModuleName s modName = S.ModuleName s (GHC.moduleNameString modName)
+
+-- | Transforms a GHC located reader name into an HST qualified name and a
+--   @Bool@ which is @True@ if the name belongs to a data constructor and
+--   @False@ otherwise.
+transformRdrName :: GHC.Located GHC.RdrName -> (S.QName GHC, Bool)
+transformRdrName (GHC.L s (GHC.Unqual name)) =
+  let s' = transformSrcSpan s
+  in  (S.UnQual s' (S.Ident s' (GHC.occNameString name)), GHC.isDataOcc name)
+transformRdrName (GHC.L s (GHC.Qual modName name)) =
+  let s' = transformSrcSpan s
+  in  ( S.Qual s'
+               (transformModuleName s' modName)
+               (S.Ident s' (GHC.occNameString name))
+      , GHC.isDataOcc name
+      )
+transformRdrName (GHC.L s (GHC.Exact name)) =
+  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), False)
+transformRdrName _ = error "Unsupported RdrName"
+
+-- | Transforms a GHC located unqualified reader name into an HST name.
+transformRdrNameUnqual :: GHC.Located GHC.RdrName -> S.Name GHC
+transformRdrNameUnqual (GHC.L s (GHC.Unqual occName)) =
+  S.Ident (transformSrcSpan s) (GHC.occNameString occName)
+transformRdrNameUnqual _ = error "Expected an unqualified name"
+
+-- | Transforms a GHC name with an HST source span into an HST special
+--   constructor.
+transformName :: S.SrcSpan GHC -> GHC.Name -> S.SpecialCon GHC
+transformName s name = case lookup name (specialConMap s) of
```

Comment:
With the patch above, unit and empty list patterns still caused the following error message in my tests.

> ```
> Internal: Expected a built-in special constructor
> CallStack (from HasCallStack):
>   error, called at src/lib/HST/Frontend/FromGHC.hs:376:10 in haskell-src-transformations-0.1.1.0-inplace:HST.Frontend.FromGHC
> ```

However, pair patterns worked correctly which suggests that the `lookup` in `specialConMap` probably does not work.

Reply:
Okay, then we should probably not use the `Eq` instance of [`Name`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/Name.html#t:Name) at all, it also has a note saying it should be used with caution.
What do you think about using `nameOccName :: Name -> OccName` and `occNameString :: OccName -> String` and implementing `specialConMap` on those strings?
Or, as this probably does retain the spelling of the built-in constructor, so `()` wouldn't be equal to `( )`, a pretty-printing function could be used instead.

Reply:
I did some more tests. As it turns out `ghc-lib` differentiates between names of type and data constructors. Unlike in `haskell-src-exts`, the list type constructor [`GHC.listTyConName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/TysWiredIn.html#v:listTyConName) is not the same as the empty list constructor [`GHC.nilDataConName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/TysWiredIn.html#v:nilDataConName), for example. Since we never need type constructors, I've changed `specialConMap` to use the data constructor names in 1af73e6c3df8cda6b987b2aa531f0254c244b52b. Furthermore, I had to change `GHC.ATyCon` to `GHC.AConLike` in `transformName` to support tuples in infix notation (e.g. `(,) x y`).